### ISSUE TITLE
Implement `stage` as optional Engine calback

### ISSUE THREAD

**Issue by sorentwo** [sorentwo/oban#770]
Staging is essential functionality and is required for Oban to operate correctly. The stager is required to process jobs and is magically injected as a plugin. It is impossible to override the stager with an alternative implementation, which is necessary before we can target alternative backends (SQLite, namely).

**Comment by rupurt** [comment]
@sorentwo I would love to run Oban on SQLite. Can you provide any links/documentation on how you require this to be implemented so I could get started on it?

**Comment by rupurt** [comment]
Oh gotcha. No worries. Looking forward to using it! Got big plans for what it enables ;)

**Comment by warmwaffles** [comment]
This would be really cool to see implemented for embedded devices.

**Comment by sorentwo** [comment]
There appears to be enough standard SQL involved in the stager that differentiating between engines isn't necessary. Closing this in favor of #819

**Comment by warmwaffles** [comment]
If you run into any issues with `ecto_sqlite3` don't hesitate to drop an issue https://github.com/elixir-sqlite/ecto_sqlite3

**Comment by sorentwo** [comment]
@warmwaffles The biggest oddity I encountered was around datetime comparison. It took a lot of debugging to figure out that without explicit `datetime(timestamp)` casts, it was making a string comparison. My workaround was a fragment that applied casting for both values: https://github.com/sorentwo/oban/pull/819/files#diff-3967bc20fbedff3a372c0ca7a3129b1f8e8894555d9b5ccc8a6a49b90b8c2b8eR251

I'm hoping something is off with my setup or that query because it's inconvenient to have a fragment for timestamp comparisons. Somehow, the time queries for staging worked without any casting 🤷

**Comment by warmwaffles** [comment]
heh this is a common problem with sqlite. There is no true datetime primitive. So if you wanted to do datetime comparisons, you have to explicitly cast it as you have done with the fragment. 😭 

If you check out this https://www.sqlite.org/quirks.html it should give you some better insight into the problem.

**Comment by warmwaffles** [comment]
I do believe we _could_ handle the datetime comparisons behind the scenes. I would just need to setup some tests to replicate what you are doing to validate.

**Comment by sorentwo** [comment]
That would be amazing. I dug through the source briefly to look for some type coercion around binary operators, but decided I would punt on it and open an issue later. Lucky me, I didn’t even have to 😁

**Comment by warmwaffles** [comment]
@sorentwo It appears to just work https://github.com/elixir-sqlite/ecto_sqlite3/commit/707fe8ec0dc9a0798f1fe8dd7533ebb1d84ae778 

Are you sure you can't do `|> where([j], j.inserted_at >= ^since))`?

**Comment by sorentwo** [comment]
@warmwaffles With that code, which is what I started with, the "scoping uniqueness by period" test fails with this error:

```
1) test insert/2 scoping uniqueness by period (Oban.Engines.Lite.Test)
     test/oban/engine_test.exs:149
     Assertion with in failed
     code:  assert job_8.id in [job_2.id, job_5.id]
     left:  7
     right: [2, 5]
     stacktrace:
       test/oban/engine_test.exs:175: (test)
```

When both sides are wrapped in `datetime` the test passes.

**Comment by warmwaffles** [comment]
What is the datatype of the `inserted_at` column and from what I can tell, `since` is a `DateTime.t()`

**Comment by sorentwo** [comment]
I can include the data later (on mobile), but can confirm now that microseconds are stored and the schema’s type has a `usec` at the end.

**Comment by warmwaffles** [comment]
@sorentwo should we take this conversation to another ticket or PR as well?

**Comment by sorentwo** [comment]
Absolutely. We’ve thoroughly hijacked this closed issue. I’ll open a new PR later this afternoon.

**Comment by jc00ke** [comment]
> Absolutely. We’ve thoroughly hijacked this closed issue. I’ll open a new PR later this afternoon.

SUBSCRIBE 🤣
